### ISSUE TITLE
Calling `read` with certain inputs causes a segfault

### ISSUE THREAD

**Issue by lexi-lambda** [racket/racket#1121]
Looks like the fix for #1118 introduces a new bug; specifically, one that causes a segmentation fault. It seems to explode when provided with a malformed pair from a nonstandard input port:

```racket
(require rnrs/io/ports-6)
(read (open-string-input-port "(#;a . b)")) ; segfault!
```

I know, a bit of a crazy combination of things to cause that particular error. It seems to work fine when given a normal Racket input port created with `open-input-string`, though.

**Comment by lexi-lambda** [comment]
After updating to the latest version and doing some more testing, I've found that I don't get this error 100% of the time, but I can get it on a pretty regular basis if I run the code from within a module. I don't *seem* to be getting the error when running at the REPL, but I was before, so who knows? When it doesn't segfault, it gives me the expected ``read: illegal use of `.'`` error.

**Comment by lexi-lambda** [comment]
Okay, narrowing things down a little, I've managed to get a segfault when executing this expression (though it seems to happen more intermittently):

```racket
(read (open-string-input-port ")"))
```

Still only seems to happen with `open-string-input-port`, not `open-input-string`.

**Comment by lexi-lambda** [comment]
It looks like my initial instinct was wrong—I think this is unrelated to 5a8d2e42041ec3e78482c50de85903c767ca0bdd—but it is still a bug. I've traced it back to [`scheme_read_err` in  `error.c`, on line 2178](https://github.com/racket/racket/blob/5cc3059de2fe4abb8a05f4f39b8b794eba9e3558/racket/src/racket/src/error.c#L2178). The `SCHEME_IPORT_NAME` macro just gets the name of a `Scheme_Input_Port` structure, but for whatever reason, this seems to sometimes be garbage, which it apparently sometimes (but only sometimes!) is in the case of R6RS's string input port.

I don't know *why* this happens, but any segfault seems like a bug, even if this is revealing a bug in r6rs-lib as well. Also, sorry for the noisy bug thread... I probably should have done more investigation myself before filing the bug report rather than just including a running tally of the information I came across. I hope the end result is helpful.
